install_requirement.py

```diff
@@ -0,0 +1,173 @@
+# this is very inefficient when only one or a few specific requirements are needed.
+#
+# Therefore, this script will evaluate the requirements defined in the project's
+# metadata and install the ones matching those being requested to be installed.
+
+from __future__ import annotations
+
+import subprocess
+import sys
+from argparse import ArgumentParser, RawDescriptionHelpFormatter
+from pathlib import Path
+from shutil import get_terminal_size
+
+from build.util import project_wheel_metadata
+from packaging.requirements import Requirement
+
+
+class RequirementsInstallerError(Exception):
+    def __init__(self, msg: str, error_no: int):
+        self.msg = msg
+        self.error_no = error_no
+
+
+class HelpText(RequirementsInstallerError):
+    """Shows script's help text."""
+
+
+class NoRequirementsFound(RequirementsInstallerError):
+    """No project requirements were found to install."""
+
+
+def parse_args():
+    width = max(min(get_terminal_size().columns, 80) - 2, 20)
+    parser = ArgumentParser(
+        description="Installs one or more PEP 517 project defined requirements",
+        formatter_class=lambda prog: RawDescriptionHelpFormatter(prog, width=width),
+    )
+    parser.add_argument(
+        "requirements",
+        type=str,
+        nargs="*",
+        help=(
+            "List of project requirements to install. Any project requirements that "
+            "start with any of these values will be installed. For instance, including "
+            "'pytest' in this list would install both pytest and pytest-xdist."
+        ),
+    )
+    parser.add_argument(
+        "--extra",
+        type=str,
+        default="",
+        help="Name of the extra where the requirements are defined",
+    )
+    parser.add_argument(
+        "--project-root",
+        type=Path,
+        default=".",
+        help=(
+            "File path to the root of the project. The current directory is used by "
+            "default."
+        ),
+    )
+
+    args = parser.parse_args()
+
+    if not args.requirements:
+        raise HelpText(parser.format_help(), error_no=-1)
+
+    return args
+
+
+def gather_requirements(
+    project_root: str | Path,
+    requested_requirements: list[str],
+    extra_name: str = "",
+) -> list[Requirement]:
+    """Identifies one or more matching requirements from a project."""
+    project_root = Path(project_root).resolve()
+    project_metadata = project_wheel_metadata(project_root, isolated=False)
+    project_requirements = [
+        requirement
+        for requirement in map(Requirement, project_metadata.get_all("Requires-Dist"))
+        if not requirement.marker or requirement.marker.evaluate({"extra": extra_name})
+    ]
+
+    matching_requirements = [
+        requirement
+        for requirement in project_requirements
+        if any(requirement.name.startswith(req) for req in requested_requirements)
+    ]
```

Comment:
Right; that is the intended use-case since many testing dependencies tend to be named with a prefix that's an implied dependency. Or in the least, that's true for the `pytest` and `coverage` plugins we're using.

This type of functionality should probably be supported with a regex but I didn't want to add that complication.

Using a literal may be more expected and allows the behavior to be more explicit.

Reply:
Yeah - this is relying on a very specific naming convention - while this may be valid for a lot of packages, there's not guarantee that it will always be true. I can also imagine a possible situation where we *literally* just want pytest and no plugins, so explicit will always be better than implicit.